### ISSUE TITLE
throughput benchmarking is misleading / not working

### ISSUE THREAD

**Issue by kdorosh** [istio/ztunnel#424]
if you make this buffer much larger (10-1000x) https://github.com/istio/ztunnel/blob/dbbeb8f341e0b2651204094c0dace53718497f2b/src/proxy.rs#L163-L164 then `cargo bench --quick` will show impossible throughput results. HBONE will be much faster than direct and tcp proxy. Thus, I have very little confidence in the throughput test as it stands today

**Comment by hzxuzhonghu** [comment]
I think it is reasonable, increasing the buffer size would improve throughput

**Comment by kdorosh** [comment]
i think it is reasonable for an increased buffer size to improve throughput

i don't think it's reasonable that increasing that buffer size can result in throughput numbers for HBONE much higher than pure TCP proxy or direct connection to upstream (no proxy at all), which is what i saw when increasing this number 10-100x

**Comment by jaellio** [comment]
@Stevenjin8

**Comment by Stevenjin8** [comment]
Pretty sure this was silently fixed in #446. I can reproduce the issue before that, but not after.
Specifically, it seems like it was the update of the `h2` crate from `0.3.15` to `0.3.16`. Here are the [release notes](https://github.com/hyperium/h2/releases) for `h2@0.3.16`. I think

> Fix panic when trying to send data and connection window is available, but stream window is not.

might be what fixed it, but I'm just speculating.

Here are the results on commit cce6f79 (right before #446):

```
throughput/direct       time:   [2.9227 ms 2.9352 ms 2.9850 ms]
                        thrpt:  [3.2715 GiB/s 3.3271 GiB/s 3.3413 GiB/s]
throughput/tcp          time:   [2.8536 ms 2.8614 ms 2.8928 ms]
                        thrpt:  [3.3759 GiB/s 3.4129 GiB/s 3.4222 GiB/s]
Benchmarking throughput/hbone
2023-07-28T19:19:33.267542Z ERROR ztunnel::proxy::inbound: hbone server copy: io error: broken pipe dur=3.634443741s
throughput/hbone        time:   [74.989 µs 76.106 µs 76.386 µs]
                        thrpt:  [127.85 GiB/s 128.32 GiB/s 130.23 GiB/s]
```
and the results after running `cargo update -p h2 --precise 0.3.16`

```
throughput/direct       time:   [2.9268 ms 2.9770 ms 2.9896 ms]
                        thrpt:  [3.2666 GiB/s 3.2803 GiB/s 3.3366 GiB/s]
throughput/tcp          time:   [2.8890 ms 2.9785 ms 3.0008 ms]
                        thrpt:  [3.2543 GiB/s 3.2787 GiB/s 3.3803 GiB/s]
throughput/hbone        time:   [16.424 ms 16.471 ms 16.661 ms]
                        thrpt:  [600.22 MiB/s 607.12 MiB/s 608.87 MiB/s]
```

The `hbone server copy error` seems to be related to #490.

Something else that I noticed is that the test will exit early on the copy error. Maybe that's related to the crazy hbone throughput numbers?

**Comment by kdorosh** [comment]
i am also able to repro https://github.com/istio/ztunnel/commit/cce6f79b77c2cbb84d3355fb24c1ff77241cbd21 but not on https://github.com/istio/ztunnel/commit/6b7dbec1500b4ac0af9fa253ac687a2053dcbd1c , so im happy to close this as fixed
